src/modules/shortcut.ts

```diff
@@ -0,0 +1,202 @@
+import {
+	command,
+	default as CookiecordClient,
+	Module,
+	listener,
+} from 'cookiecord';
+import { Message, MessageEmbed, User } from 'discord.js';
+import { send } from 'process';
+import { BaseEntity, DeepPartial, ObjectType } from 'typeorm';
+import { Shortcut } from '../entities/Shortcut';
+import { BLOCKQUOTE_GREY, BLOCKQUOTE_GREY_NUMBER, TS_BLUE } from '../env';
+import { sendWithMessageOwnership } from '../util/send';
+
+// https://stackoverflow.com/a/3809435
+const LINK_REGEX = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)/;
+
+export class ShortcutModule extends Module {
+	constructor(client: CookiecordClient) {
+		super(client);
+	}
+
+	@listener({ event: 'message' })
+	async runShortcut(msg: Message) {
+		const { content } = msg;
+		const commandPart = content.split(' ')[0];
+		if (!commandPart.startsWith('!')) return;
+		const id = commandPart.slice(1);
+		if (this.client.commandManager.getByTrigger(id)) return;
+		const shortcut = await this.getShortcut(id);
+		if (!shortcut) return;
+		if (shortcut.content)
+			return await sendWithMessageOwnership(msg, shortcut.content);
+		const owner = await this.client.users.fetch(shortcut.owner);
+		const embed = new MessageEmbed({
+			...shortcut,
+			image: undefined,
+		});
+		embed.setAuthor(owner.tag, owner.displayAvatarURL());
+		if (shortcut.image) embed.setImage(shortcut.image);
+		await sendWithMessageOwnership(msg, { embed });
+		await Shortcut.createQueryBuilder()
+			.update()
+			.set({ uses: () => 'uses + 1' })
+			.execute();
+	}
+
+	@command({
+		single: true,
+		description: 'Shortcut: Create or edit a shortcut',
+	})
+	async shortcut(msg: Message, arg: string) {
+		const [name, ...titleParts] = arg.split(' ');
+		const title = titleParts?.join(' ') ?? '';
+
+		if (!name) {
+			return await sendWithMessageOwnership(
+				msg,
+				':x: You have to supply a name for the command',
+			);
+		}
+
+		const id = this.getId(msg.author, name);
+		const existingShortcut = await this.getShortcut(id);
+
+		if (existingShortcut && existingShortcut.owner !== msg.author.id)
+			return await sendWithMessageOwnership(
+				msg,
+				":x: Cannot edit another's shortcut",
+			);
+
+		const referencedMessage = await msg.channel.messages.fetch(
+			msg.reference?.messageID!,
+		);
+		if (!referencedMessage)
+			return await sendWithMessageOwnership(
+				msg,
+				':x: You have to reply to a comment to make it a shortcut',
+			);
+
+		const description = referencedMessage.content;
+		const referencedEmbed = referencedMessage.embeds[0];
+		const base = {
+			id,
+			uses: existingShortcut?.uses ?? 0,
+			owner: msg.author.id,
+		};
+
+		let data: Omit<Shortcut, keyof BaseEntity> | undefined;
+		if (LINK_REGEX.exec(description)?.[0] === description)
```

Comment:
You might be able to get away with using [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) instead.

Reply:
`new URL("https://a.b.c/d a")` succeeds, so I don't think that will work.

Reply:
This isn't too big of a deal but for the sake of completeness, it seems `URL` transforms that into a valid URL:

```
> new URL("https://a.b.c/d a")
URL {
  href: 'https://a.b.c/d%20a',
  host: 'a.b.c',
[cut]
  pathname: '/d%20a',
  search: '',
}
> new URL("there's no way this is valid")
Uncaught TypeError [ERR_INVALID_URL]: Invalid URL: there's no way this is valid
    at onParseError (internal/url.js:259:9)
 ```